common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1457,87 +1459,91 @@ public SSLEngineResult wrap(ByteBuffer[] srcs, int srcsOffset, int srcsLength, B
+                            return pendingNetResult != null ? pendingNetResult
+                                    : CLOSED_NOT_HANDSHAKING;
+                        case SSL_ERROR_WANT_READ:
+                            // If there is no pending data to read from BIO we should go back to
+                            // event loop and try
+                            // to read more data [1]. It is also possible that event loop will
+                            // detect the socket
+                            // has been closed. [1]
+                            // https://www.openssl.org/docs/manmaster/man3/SSL_write.html
+                            pendingNetResult = readPendingBytesFromBIO(
+                                    dst, bytesConsumed, bytesProduced, handshakeStatus);
+                            return pendingNetResult != null
+                                    ? pendingNetResult
+                                    : new SSLEngineResult(getEngineStatus(), NEED_UNWRAP,
+                                    bytesConsumed, bytesProduced);
+                        case SSL_ERROR_WANT_WRITE:
+                            // SSL_ERROR_WANT_WRITE typically means that the underlying
+                            // transport is not writable
+                            // and we should set the "want write" flag on the selector and try
+                            // again when the
+                            // underlying transport is writable [1]. However we are not directly
+                            // writing to the
+                            // underlying transport and instead writing to a BIO buffer. The
+                            // OpenSsl documentation
+                            // says we should do the following [1]:
+                            //
+                            // "When using a buffering BIO, like a BIO pair, data must be
+                            // written into or retrieved
+                            // out of the BIO before being able to continue."
+                            //
+                            // So we attempt to drain the BIO buffer below, but if there is no
+                            // data this condition
+                            // is undefined and we assume their is a fatal error with the
+                            // openssl engine and close.
+                            // [1] https://www.openssl.org/docs/manmaster/man3/SSL_write.html
+                            pendingNetResult = readPendingBytesFromBIO(
+                                    dst, bytesConsumed, bytesProduced, handshakeStatus);
+                            return pendingNetResult != null ? pendingNetResult
+                                    : NEED_WRAP_CLOSED;
+                        default:
+                            // Everything else is considered as error
+                            closeAll();
+                            throw newSslExceptionWithMessage("SSL_write");
```

Comment:
nit: will sslError be useful in error message?

Reply:
Added it... In theory this should never happen though